### PR TITLE
Toast focus management take 2

### PR DESCRIPTION
**Pull request by snowystinger**
Closes 

In association with https://github.com/adobe/react-spectrum/pull/6011

There are still some known issues, they are listed in our GA ticket https://github.com/orgs/adobe/projects/19/views/4?pane=issue&itemId=55618129

## ✅ Pull Request Checklist:

- [ ] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/WAI/ARIA/apg/)

## 📝 Test Instructions:

Test instructions in https://github.com/adobe/react-spectrum/pull/6011 description

### CODE REVIEW

File: packages/@react-aria/toast/src/useToastRegion.ts [hunk 1]
```diff
@@ -29,14 +29,51 @@ export function useToastRegion<T>(props: AriaToastRegionProps, state: ToastState
...
+      .filter(((t) => !state.visibleToasts.some(t2 => t.key === t2.key)));
+
+    // If the focused toast was removed, focus the next or previous toast.
+    if (removedToasts.some(t => t.i === focusedToast.current)) {
+      let nextToast = focusedToast.current;
+      if (nextToast >= toasts.current.length) {
+        nextToast = toasts.current.length - 1;
+      }
+      if (nextToast !== -1) {
+        focusWithoutScrolling(toasts.current[nextToast]);
```
Comment (majornista) [inline packages/@react-aria/toast/src/useToastRegion.ts:70]: We may want to wait a frame before restoring focus here, so that the list item count does not still include the item being removed.

**Comment by majornista** [comment]
@snowystinger With recent changes, focus is getting lost to the document.body when the top toast, the last item in the list gets removed and other toasts remain.

**Comment by snowystinger** [comment]
> @snowystinger With recent changes, focus is getting lost to the document.body when the top toast, the last item in the list gets removed and other toasts remain.

thanks for the identification and fix, I've added a test case to cover the behavior

**Review by majornista — APPROVED** [review]
I tested this change using NVDA in Firefox, Chrome and Edge, and the Toast messages are being announced.

File: packages/@react-aria/toast/src/useToastRegion.ts [hunk 2]
```diff
@@ -29,14 +29,79 @@ export function useToastRegion<T>(props: AriaToastRegionProps, state: ToastState
...
+        }
+        i++;
+      }
+      while (i < removedToasts.length) {
+        if (!removedToasts[i].isRemoved) {
+          nextToast = i - 1;
+          break;
+        }
+        i++;
+      }
```
Comment (LFDanLu) [inline packages/@react-aria/toast/src/useToastRegion.ts:88]: bit of a nit, but I found this a bit hard to follow, could we instead just slice the removedToasts list into two (one going from the 0 -> removedToastIndex and the other going from removedToastIndex -> end) and use findLastIndex with the same `!removedToasts.isRemoved` condition on the first slice to get the index of the prevToast and findIndex + the removedToastIndex on the second slice to get the nextToast index?

Thinking on this, it might just be the "removedToasts" naming that is tripping me up since that list contains all the toasts but just adds extra data to mark the removed ones.

File: packages/@react-spectrum/text/src/Text.tsx [hunk 3]
```diff
@@ -26,7 +26,7 @@ function Text(props: TextProps, ref: DOMRef) {
   let domRef = useDOMRef(ref);
 
   return (
-    <span {...filterDOMProps(otherProps)} {...styleProps} ref={domRef}>
+    <span role="none" {...filterDOMProps(otherProps)} {...styleProps} ref={domRef}>
```
[the remark this answers is not in the document]
Comment (LFDanLu) [reply packages/@react-spectrum/text/src/Text.tsx:29]: Ah I see, the Button for the toast action. Will have to double check everywhere we use Text to see if there are any changes to the screen reader performance

File: packages/@react-aria/toast/src/useToast.ts [hunk 4]
```diff
@@ -39,6 +41,7 @@ export interface ToastAria {
  * Provides the behavior and accessibility implementation for a toast component.
  * Toasts display brief, temporary notifications of actions, errors, or other events in an application.
  */
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
```
[the remark this answers is not in the document]
Comment (snowystinger) [reply packages/@react-aria/toast/src/useToast.ts:44]: it's more of a, we've needed to add it back in later in other cases. I'd prefer to know we'll have access to it

File: packages/@react-aria/toast/src/useToast.ts [hunk 5]
```diff
@@ -58,44 +61,35 @@ export function useToast<T>(props: AriaToastProps<T>, state: ToastState<T>, ref:
...
 
   let titleId = useId();
   let descriptionId = useSlotId();
   let stringFormatter = useLocalizedStringFormatter(intlMessages, '@react-aria/toast');
 
   return {
     toastProps: {
-      role: 'alert',
+      role: 'alertdialog',
+      'aria-modal': 'false',
```
[the remark this answers is not in the document]
Comment (snowystinger) [reply packages/@react-aria/toast/src/useToast.ts:78]: @majornista correct me if I'm wrong, but this is just to possibly forward proof? since alertdialog would be considered a modal by default, and so we use aria-modal to communicate that it's not, in fact, a modal

File: packages/@react-spectrum/text/src/Text.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (snowystinger) [reply packages/@react-spectrum/text/src/Text.tsx:29]: yeah, i haven't noticed anything yet

File: packages/@react-aria/toast/src/useToastRegion.ts (hunk 2 above)
Comment (snowystinger) [reply packages/@react-aria/toast/src/useToastRegion.ts]: yeah, it was hard to write too
I've clarified with renames and some more comments

File: packages/@react-aria/toast/src/useToast.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (majornista) [reply packages/@react-aria/toast/src/useToast.ts:78]: `aria-modal="false"` doesn't really make a difference for screen reader behavior; the default is for a dialog to be non-modal. However, the [APG design pattern](https://www.w3.org/WAI/ARIA/apg/patterns/alertdialog/) describes an `alertdialog` as modal, so I thought it best to be explicit:

> An alert dialog is a [modal dialog](https://www.w3.org/WAI/ARIA/apg/patterns/dialog-modal/) that interrupts the user's workflow to communicate an important message and acquire a response. Examples include action confirmation prompts and error message confirmations. The [alertdialog](https://w3c.github.io/aria/#alertdialog) role enables assistive technologies and browsers to distinguish alert dialogs from other dialogs so they have the option of giving alert dialogs special treatment, such as playing a system alert sound.

I use `role="alertdialog"`, as a container for the toast message content, which has `role="alert"` to announce as a live region, and the button to dismiss the Toast, for a few reasons.

1. A Toast is kind of a tiny non-modal dialog, so it didn't seem like much of a stretch.
2. Using a dialog as a container helps disambiguate between the Dismiss buttons when more than one Toast is displayed.
3. When we close a Toast, unlike the Dismiss button for an adjacent Toast, the dialog provides a uniquely labeled widget element to receive focus, and is easy to style using the React-Spectrum focus ring.
4. Depending on the assistive technology being used, an alert dialog will open with an earcon, which I see as nice feature to distinguish a Toast from other content the screen reader may be announcing.

File: packages/@react-spectrum/text/src/Text.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (majornista) [reply packages/@react-spectrum/text/src/Text.tsx:29]: Yeah, this was something I noticed in testing with VoiceOver. When a Toast included an action button, the button text was announced twice. This has to do with how live regions work, announcing both element "additions" and "text." VoiceOver seem to interpret the Text component's `span` element child of the Button as an addition, and announces the text more than once.

File: packages/@react-aria/toast/src/useToast.ts [hunk 6]
```diff
@@ -25,8 +25,10 @@ export interface AriaToastProps<T> extends AriaLabelingProps {
 }
 
 export interface ToastAria {
-  /** Props for the toast container element. */
+  /** Props for the toast container, non-modal dialog element. */
   toastProps: DOMAttributes,
+  /** Props for the toast content alert message. */
+  contentProps: DOMAttributes,
```
Comment (devongovett) [inline packages/@react-aria/toast/src/useToast.ts:31]: Do we still need both contentProps and titleProps as separate elements or are they the same thing now?

File: packages/@react-aria/toast/src/useToastRegion.ts [hunk 7]
```diff
@@ -29,14 +29,80 @@ export function useToastRegion<T>(props: AriaToastRegionProps, state: ToastState
...
+      // prioritize going to newer toasts
+      if (prevToast >= 0 && prevToast < toasts.current.length) {
+        focusWithoutScrolling(toasts.current[prevToast]);
+      } else if (nextToast >= 0 && nextToast < toasts.current.length) {
+        focusWithoutScrolling(toasts.current[nextToast]);
+      }
+    }
+
+    prevVisibleToasts.current = state.visibleToasts;
+  }, [state.visibleToasts, ref]);
```
Comment (devongovett) [inline packages/@react-aria/toast/src/useToastRegion.ts:104]: This logic is quite similar to what we have for other collection components when removing rows. Wonder if we could centralize this at some point.

**Review by devongovett — APPROVED** [review]
Do we need to update the aria docs example?

File: packages/@react-aria/toast/src/useToast.ts (hunk 6 above)
Comment (snowystinger) [reply packages/@react-aria/toast/src/useToast.ts:31]: yeah, it still provides a useful distinction for the aria-labelledby. You can put things like the icon on the outside of the title so it isn't read off all the time

**Comment by snowystinger** [comment]
Yeah, we should update the hooks docs, followup https://github.com/orgs/adobe/projects/19/views/4?pane=issue&itemId=65286699
Starting branch here: https://github.com/adobe/react-spectrum/pull/6468

File: packages/@react-aria/toast/src/useToast.ts (hunk 6 above)
Comment (levrik) [reply packages/@react-aria/toast/src/useToast.ts:31]: Is the new `contentProps` required to be used? The documentation isn't really clear about this. I currently don't have a wrapper element around my toast title and description as they can be arranged differently in layout with CSS Grid.

Comment (snowystinger) [reply packages/@react-aria/toast/src/useToast.ts:31]: It probably depends a bit on how complex your layout is. It can be helpful to target the live announcer to a more succinct announcement. So I'd say good practice, but possibly not strictly necessary. You'll have to test it out with a screen reader to see if not including it is sufficient.

Comment (levrik) [reply packages/@react-aria/toast/src/useToast.ts:31]: @snowystinger I see. Thanks for the response!

Comment (majornista) [reply packages/@react-aria/toast/src/useToast.ts:31]: @levrik and @snowystinger The new `contentProps` includes the `role="alert"` attribute, which prompts live region announcement when the Toast appears, so I would say these props are necessary.
